common/src/main/java/io/netty/util/concurrent/SingleThreadEventExecutor.java

```diff
@@ -363,22 +419,39 @@ protected boolean runAllTasks(long timeoutNanos) {
             if ((runTasks & 0x3F) == 0) {
                 lastExecutionTime = ScheduledFutureTask.nanoTime();
                 if (lastExecutionTime >= deadline) {
-                    break;
+                    return DeadlineRunStatus.TIMED_OUT;
                 }
             }
 
-            task = pollTask();
+            task = pollTaskFrom(taskQueue);
             if (task == null) {
                 lastExecutionTime = ScheduledFutureTask.nanoTime();
                 break;
             }
         }
+        lastExecutionTimeCallback.call(lastExecutionTime);
+        return DeadlineRunStatus.RAN_AT_LEAST_ONCE;
+    }
 
-        this.lastExecutionTime = lastExecutionTime;
-        return true;
+    protected enum DeadlineRunStatus {
+        NO_TASKS_FOUND,
+        TIMED_OUT,
+        RAN_AT_LEAST_ONCE
+    }
+
+    protected interface Action<T> {
```

Comment:
Why this needs to be generic ? Can we just not make it generic and use something like `call(long ...)`. This way you not need boxing etc.